lib/server.js

```diff
@@ -92,459 +92,114 @@ function Server(originalCommand, baseDir, listenPort, controlPath) {
-        // uncertain, don't rely on it.
-        return asyncCb();
-      }
-      m.processId = proc.id;
-      m.workerId = wid;
-      m.timeStamp = req.metrics.timestamp;
-      Metric.upsert(m, asyncCb);
-    });
-  }
-
-  function deleteOld() {
-    var now = new Date().getTime();
-    var where = {
-      timeStamp: {lt: now - 5 * 60 * 1000},
-    };
-    Metric.destroyAll(where, function(err) {
-      assert.ifError(err);
-      if (callback) return callback(err);
-    });
-  }
-};
-
-Server.prototype._onExpressUsageRecord = function(req, callback) {
-  debug('express usage record: %j', req.record);
-  var Process = this._app.models.ServiceProcess;
-  var ExpressUsageRecord = this._app.models.ExpressUsageRecord;
-  var record = req.record;
-  var pid = +record.process.pid;
-
-  var q = {where: {
-    serviceInstanceId: 1,
-    pid: pid,
-    stopTime: null
-  }};
-
-  Process.findOne(q, function(err, proc) {
-    assert.ifError(err);
-    if (!proc) {
-      debug('Ignoring metrics for unknown worker pid: %d', pid);
-      return deleteOld();
-    }
+Server.prototype._onMasterStart =
+  function _onMasterStart(supervisorInfo, callback) {
+    var self = this;
+    var current = runner.current();
+    var commit = current.commit;
+    var strongPmInfo = require(path.join(__dirname, '..', 'package.json'));
 
-    var usageRecord = {
-      processId: pid,
-      workerId: proc.workerId,
-      timestamp: record.timestamp,
-      usage: record
-    };
+    current.request({cmd: 'status'}, function(status) {
+      if (!self._isStarted) return;
+      debug('on running: %j', commit);
+
+      var supervisorPid = supervisorInfo.pid;
+      current.once('exit', function(status) {
+        // Sometimes the supervisor process may exit without any events, so we
+        // must record its and its childrens death
+        self._onRunnerRequest({
+          cmd: 'exit', id: 0, pid: supervisorPid, reason: status
+        });
+      });
 
-    ExpressUsageRecord.create(usageRecord, function(err) {
-      assert.ifError(err);
-      deleteOld();
-    });
-  });
+      var req = {cmd: 'npm-ls'};
+      current.request(req, function(appInfo) {
+        if (appInfo.error) {
+          return callback(Error(appInfo.error));
+        }
 
-  function deleteOld() {
-    var now = new Date().getTime();
-    var where = {timestamp: {lt: now - 5 * 60 * 1000}};
-    ExpressUsageRecord.destroyAll(where, function(err) {
-      assert.ifError(err);
-      callback();
+        // Process information
+        supervisorInfo.id = 0; // Supervisor is worker id 0
+        supervisorInfo.ppid = process.pid;
+
+        // Instance information
+        supervisorInfo.commitHash = commit ? commit.hash : undefined;
+        supervisorInfo.npmModules = appInfo;
+        supervisorInfo.PMPort = self._listenPort;
+        supervisorInfo.containerVersionInfo = {
+          os: {
+            platform: os.platform(),
```

Comment:
Would you mind putting a `TODO` comment in here to make sure we don't miss then when we refactor the runner into drivers? Containerized pm vs. containerized app comes to mind as a case where this code will break as is.